api/src/main/java/com/google/sps/SendNotifications.java

```diff
@@ -20,98 +16,156 @@
+      throws IOException {
 
-    System.out.println("Google Credentials created...");
-
-    String projectId = "phoebeliang-step";
-    FirebaseOptions options = new FirebaseOptions.Builder().setCredentials(credentials).setProjectId(projectId).build();
-    FirebaseApp.initializeApp(options);
-
-    System.out.println("App Initialized...");
-
-    Firestore db = FirestoreClient.getFirestore();
-
-    String gameID = "dT6U1wQKR1rGJcEncyMS";
-    List<Email> emails = new ArrayList<>();
+    String gameID = request.getParameter("gameID");
+    String emailType = request.getParameter("emailType");
+    List<EmailCreation> emails = new ArrayList<>();
 
     DocumentReference game = db.collection("games").document(gameID);
-
+    CollectionReference players = db.collection("users");
     try {
       DocumentSnapshot gameDocSnap = game.get().get();
       List<String> playerNames = (List<String>) gameDocSnap.get("players");
 
-      for(String player: playerNames) {
-        ApiFuture<QuerySnapshot> futurePlayers = db.collection("users").whereEqualTo("username", player).get();
-        List<QueryDocumentSnapshot> usersDocSnap = futurePlayers.get().getDocuments();
+      for (String player : playerNames) {
+        DocumentSnapshot usersDocSnap = players.document(player).get().get();
 
         // Retrieves player information
-        Map<String,Object> playerInfo = usersDocSnap.get(0).getData();
-        String playerEmail = (String) playerInfo.get("email");
-        String playerName = (String) playerInfo.get("username");
-
-        System.out.println(playerEmail);
+        String playerEmail = usersDocSnap.getString("email");
+        String playerName = usersDocSnap.getString("username");
 
         // Adding player to Email object
-        emails.add(new Email(new User(playerEmail, playerName)));
+        if (emailType.equalsIgnoreCase("start")) {
+          emails.add(EmailCreation.startGameEmail(gameID, new User(playerEmail, playerName)));
+        } else if (emailType.equalsIgnoreCase("end") || emailType.equalsIgnoreCase("turn")) {
+          emails.add(EmailCreation.endGameEmail(gameID, new User(playerEmail, playerName)));
+        }
 
-        // Adds Start game message to the email object
-        if(emails.isEmpty()) {
+        // Checks if there are any players
+        if (emails.isEmpty()) {
           throw new IllegalArgumentException();
-        } else {
-          emails.get(emails.size()-1).playerTurn(gameID);
         }
       }
     } catch (Exception e) {
-      System.out.println(e);
+      System.out.println("Exception in gatherRecipients" + e);
     }
 
     return emails;
   }
 
-  @GetMapping("/notifyTurn")
-  public void playerTurnEmail() throws IOException {
-    List<Email> emails = gatherRecipients();
+  public EmailCreation getNextPlayer(HttpServletRequest request, HttpServletResponse response, Firestore db)
```

Comment:
What's the reasoning behind passing around an instance of firestore as opposed to just using getInstance()?

Reply:
Did you mean the getFirestore() method? I couldn't find a getInstance() method. I mistakenly associated the getFirestore() with the initialization of the app originally. Something I just thought is the fact that we'd be creating multiple Firestore objects rather than just using one throughout the program. What's your thoughts?

Reply:
yeah, there is actually an error if we try to initialize multiple {default} instances, which is why i had to add firebase.java into a separate file to check initialization.

Reply:
Right, but the creating multiple **Firestore** db object isn't a problem? I know it probably doesn't take up that much memory, but whats the difference between passing the instance around rather than making multiple? just curious

Reply:
[getFirestore() is static](https://firebase.google.com/docs/reference/admin/java/reference/com/google/firebase/cloud/FirestoreClient), not multiple instances created.
This is a design decision around the intention of the API. I think no 'correct' answer, as long as there is solid reasoning behind it.

Reply:
okay, i went with the doing the getFirestore in each method. i pushed the new changes!